### PR TITLE
Furnace based machines could no longer be used as vanilla furnace with the hoppers.

### LINKED ISSUE #2449: Furnace based machines can be used as vanilla furnace with the hoppers
## :round_pushpin: Description (REQUIRED)

Furnace based machines can be used as vanilla furnace with the hoppers.

## :bookmark_tabs: Steps to reproduce the Issue (REQUIRED)

1. Build this simple furnace system from hopper and chests.

2. The furnace is Electric ore grinder.

3. Place gold ore towards the chest.
4. Place coal next to the furnace chest.
5. Open the chest under the furnace.

Capture

## :bulb: Expected behavior (REQUIRED)

Furnace based machines not used as vanilla furnace with the hoppers.

## :open_file_folder: /error-reports/ Folder

No error report.

## :compass: Environment (REQUIRED)

[15:59:52 INFO]: This Server uses the following setup of Slimefun:
[15:59:52 INFO]: Paper v1_16_R2
[15:59:52 INFO]: CS-CoreLib vDEV - 91 (git e42a05d1)
[15:59:52 INFO]: Slimefun vDEV - 712 (git 15884c65)
[15:59:52 INFO]: Metrics build: #17
[15:59:52 INFO]:
[15:59:52 INFO]: Installed Addons: (0)

### LINKED ISSUE #2511: Smoker based machines can be used as vanilla smoker with hoppers
## :round_pushpin: Description (REQUIRED)

Auto drier and Auto brewer can be used as vanilla smoker with hoppers.

## :bookmark_tabs: Steps to reproduce the Issue (REQUIRED)

1. Build this simple smoker system from hopper and chests.

2. Place Raw Porkchop towards the chest.
3. Place coal next to the furnace chest.
4. Open the chest under the furnace.

Capture

## :bulb: Expected behavior (REQUIRED)

Smoker based machines not used as vanilla smoker with the hoppers.

## :open_file_folder: /error-reports/ Folder

No error report.

## :compass: Environment (REQUIRED)

[18:54:57 INFO]: Paper v1_16_R2
[18:54:57 INFO]: CS-CoreLib vDEV - 91 (git e42a05d1)
[18:54:57 INFO]: Slimefun vDEV - 729 (git 38bf2280)
[18:54:57 INFO]: Metrics build: #17
[18:54:57 INFO]: Backwards compatibility enabled!
[18:54:57 INFO]:
[18:54:57 INFO]: Installed Addons: (10)
[18:54:57 INFO]:  DyedBackpacks vDEV - 9 (git c42435c1)
[18:54:57 INFO]:  ExtraTools vDEV - 20 (git cfc5b167)
[18:54:57 INFO]:  PomaExpansion vDEV - 10 (git 3f46213c)
[18:54:57 INFO]:  LiteXpansion vDEV - 74 (git f803f67d)
[18:54:57 INFO]:  SlimefunOreChunks vDEV - 9 (git 88dd72b0)
[18:54:57 INFO]:  EcoPower vDEV - 2 (git b0c0c875)
[18:54:57 INFO]:  HotbarPets vDEV - 31 (git 5909202f)
[18:54:57 INFO]:  ExtraGear vDEV - 12 (git 4570ea2c)
[18:54:57 INFO]:  SlimefunLuckyBlocks vDEV - 29 (git db79cb7c)
[18:54:57 INFO]:  FluffyMachines vDEV - 30 (git fca125f9)

### PR DESCRIPTION
**Pull request by CURVX**
## Description

Added `HopperListener` and an interface `NotHopperable`

## Changes

- Added EventHandler `onHopperInsert` to cancel `InventoryMoveItemEvent` from hopper to issue specific Furnace/Smoker based Slimefun machines.
- `ElectricFurnace`, `AutoBrewer`, `AutoDrier`, `ElectricOreGrinder`, `ElectricIngotPulverizer` & `ElectricSmeltery` now implements `NotHopperable`.

## Related Issues

Resolves #2449 
Resolves #2511 

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunPlugin.java [hunk 1]
```diff
@@ -13,6 +13,7 @@
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
+import io.github.thebusybiscuit.slimefun4.implementation.listeners.*;
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunPlugin.java:16]: never ever static import

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ItemMoveListener.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    public ItemMoveListener(@Nonnull SlimefunPlugin plugin) {
+        plugin.getServer().getPluginManager().registerEvents(this, plugin);
+    }
+
+    @EventHandler
+    public void onItemMove(InventoryMoveItemEvent e) {
+        Location loc = e.getDestination().getLocation();
+        if (loc != null && BlockStorage.hasBlockInfo(loc)) {
+            if (BlockStorage.check(loc, SlimefunItems.ELECTRIC_ORE_GRINDER.getItemId())) {
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ItemMoveListener.java:26]: not sure if this adds all furnace type machines but for slimefun it self you forgot Electric furnace and Electric Smeltery
it would be nice if this blocks it for all Furnace type machines even for addons.

**Comment by poma123** [comment]
Just a small comment:
I'd rather move this to VanillaMachinesListener. Also, I think we should listen to all Slimefun blocks (or all furnace typed Slimefun blocks), not only the ore grinder.

**Comment by J3fftw1** [comment]
> 
> 
> Just a small comment:
> I'd rather move this to VanillaMachinesListener. Also, I think we should listen to all Slimefun blocks (or all furnace typed Slimefun blocks), not only the ore grinder.

Dont forget to make an exceptions for Furnaces without custom inventories like the enhanced furnaces

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunPlugin.java (hunk 1 above)
Comment (CURVX) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunPlugin.java:16]: image
From now NEVER! :D

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ItemMoveListener.java (hunk 2 above)
Comment (CURVX) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/ItemMoveListener.java:26]: Yes, I will be adding the rest and excluding the Enchanced Furnaces. I don't know how to though without making it messy. Will look through the codebase to get an idea.

**Comment by CURVX** [comment]
> 
> 
> Just a small comment:
> I'd rather move this to VanillaMachinesListener. Also, I think we should listen to all Slimefun blocks (or all furnace typed Slimefun blocks), not only the ore grinder.

I couldn't find the file you mentioned. Or are you saying to rename the filename?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunPlugin.java (hunk 1 above)
Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/SlimefunPlugin.java:16]: :PanLove:

**Comment by J3fftw1** [comment]
i think he wants you to create the file

**Comment by CURVX** [comment]
> I couldn't find the file you mentioned. Or are you saying to rename the filename?

I looked through the commit history, the file got moved to `listeners -> crafting` disintegrating into small ones in 8440143f416791c4817e42062577501f12d57796.

**Comment by J3fftw1** [comment]
then you would need to make FurnaceListener

**Comment by CURVX** [comment]
This checks for a Slimefun furnace containing "Electric" in its Inventory name and cancels the event. Will this approach work? If not, how do approach it without checking like 10-12 furnace based sf machines. Or "live with it"? :D

**Comment by TheBusyBiscuit** [comment]
> This checks for a Slimefun furnace containing "Electric" in its Inventory name and cancels the event. Will this approach work? If not, how do I approach it without checking like 16-20 furnace based sf machines. Or "live with it"? :D

Ugh, that is a really bad way to about this.
Theoretically you could just rename your furnace in an anvil to brick it.
This would also cause potentially countless issues with other plugins and it would also be another hardcoded localization trait, which would set us back on the way towards item localization.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/FurnaceListener.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    public FurnaceListener(@Nonnull SlimefunPlugin plugin) {
+        plugin.getServer().getPluginManager().registerEvents(this, plugin);
+    }
+
+    @EventHandler
+    public void onFurnace(InventoryMoveItemEvent e) {
+        Location loc = e.getDestination().getLocation();
+        if (loc != null && BlockStorage.hasBlockInfo(loc) && e.getSource().getType() == InventoryType.HOPPER) {
+            boolean isElectricOreGrinder = BlockStorage.check(loc, SlimefunItems.ELECTRIC_ORE_GRINDER.getItemId()) || BlockStorage.check(loc, SlimefunItems.ELECTRIC_ORE_GRINDER_2.getItemId());
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/FurnaceListener.java:24]: Yes thjs would work but wouldn't it be really tedious to keep adding machines to this that are also furnaces i think the best way to do this is to check if its a furnace and it has an id then cancel the event

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/FurnaceListener.java (hunk 3 above)
Comment (CURVX) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/FurnaceListener.java:24]: Now that you pointed it out, yes this would be tedious to manage. The way you suggested, won't this affect enhanced, reinforced and carbonado furnaces? Also, won't this cause issue for other furnace based addon machines which wants to retain this "cooking/smelting" functionality?

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/FurnaceListener.java:24]: Then maybe some sort of implementation if you want it toggled on or off 
Then people can decide thereself?
Just throwing in ideas

Comment (CURVX) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/FurnaceListener.java:24]: How about an abstract `class SlimefunContainer` extending `AContainer` and then extending the `SlimefunContainer` in all those furnace based machines which extends `AContainer` and the listener cancels the `InventoryMoveItemEvent` from hopper if the destination block is an instance of `SlimefunContainer`? 

**TL;DR:** Like Fluffy did in commit https://github.com/NCBPFluffyBear/FluffyMachines/commit/52b6fec3b64e62498f41ee15b61de5385145466b

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/FurnaceListener.java]: That could work aswell

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import io.github.thebusybiscuit.slimefun4.implementation.SlimefunPlugin;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.abstractItems.VanillaContainer;
+import me.mrCookieSlime.Slimefun.api.BlockStorage;
+
+/**
+ * This {@link Listener} prevents item from being transferred to a
+ * {@link VanillaContainer} using a hopper.
+ *
+ */
+
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java:20]: Remove this empty line.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import org.bukkit.event.inventory.InventoryType;
+
+import io.github.thebusybiscuit.slimefun4.implementation.SlimefunPlugin;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.abstractItems.VanillaContainer;
+import me.mrCookieSlime.Slimefun.api.BlockStorage;
+
+/**
+ * This {@link Listener} prevents item from being transferred to a
+ * {@link VanillaContainer} using a hopper.
+ *
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java:18]: Missing `@author` tag.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    public HopperListener(@Nonnull SlimefunPlugin plugin) {
+        plugin.getServer().getPluginManager().registerEvents(this, plugin);
+    }
+
+    @EventHandler
+    public void fromHopper(InventoryMoveItemEvent e) {
+        Location loc = e.getDestination().getLocation();
+        if (e.getSource().getType() == InventoryType.HOPPER && loc != null && BlockStorage.hasBlockInfo(loc)) {
+            if (BlockStorage.check(loc.getBlock()) instanceof VanillaContainer) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java:31]:
```suggestion
        if (e.getSource().getType() == InventoryType.HOPPER && loc != null && BlockStorage.hasBlockInfo(loc)) {
            if (BlockStorage.check(loc.getBlock()) instanceof VanillaContainer) {
```
No need to nest this here, connect the two with &&.
Also, it would be redundant to do `hasBlockInfo` and `check()`.
The check value on its own will work already since it would be null otherwise.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+public class HopperListener implements SlimefunCraftingListener {
+
+    public HopperListener(@Nonnull SlimefunPlugin plugin) {
+        plugin.getServer().getPluginManager().registerEvents(this, plugin);
+    }
+
+    @EventHandler
+    public void fromHopper(InventoryMoveItemEvent e) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java:28]: Please follow the method naming convention, `onHopperInsert` for example.

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/abstractItems/VanillaContainer.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package me.mrCookieSlime.Slimefun.Objects.SlimefunItem.abstractItems;
```
Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/abstractItems/VanillaContainer.java:1]: Do **not** use this package.

File: src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/abstractItems/VanillaContainer.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import org.bukkit.inventory.ItemStack;
+
+import me.mrCookieSlime.Slimefun.Lists.RecipeType;
+import me.mrCookieSlime.Slimefun.Objects.Category;
+import me.mrCookieSlime.Slimefun.api.SlimefunItemStack;
+
+public abstract class VanillaContainer extends AContainer {
```
Comment (TheBusyBiscuit) [inline src/main/java/me/mrCookieSlime/Slimefun/Objects/SlimefunItem/abstractItems/VanillaContainer.java:11]: Missing javadocs, also not the best name. The name could imply different stuff and it's even worse without the docs.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import me.mrCookieSlime.Slimefun.api.BlockStorage;
+
+/**
+ * This {@link Listener} prevents item from being transferred to a
+ * {@link AbstractVanillaBlockInventory} using a hopper.
+ *
+ * @author CURVX
+ *
+ */
+public class HopperListener implements SlimefunCraftingListener {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java:22]: Why does this implement the SlimefunCraftingListener...?
It doesn't have anything to do with crafting...?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+    public HopperListener(@Nonnull SlimefunPlugin plugin) {
+        plugin.getServer().getPluginManager().registerEvents(this, plugin);
+    }
+
+    @EventHandler
+    public void onHopperInsert(InventoryMoveItemEvent e) {
+        Location loc = e.getDestination().getLocation();
+        if (e.getSource().getType() == InventoryType.HOPPER && loc != null && BlockStorage.check(loc.getBlock()) instanceof AbstractVanillaBlockInventory) {
+                e.setCancelled(true);
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java:32]: indentation

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
+package io.github.thebusybiscuit.slimefun4.implementation.listeners.crafting;
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java:1]: Why the crafting package?

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/inventory/AbstractVanillaBlockInventory.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+ * @see ElectricSmeltery
+ *
+ */
+public abstract class AbstractVanillaBlockInventory extends AContainer {
+
+    @ParametersAreNonnullByDefault
+    public AbstractVanillaBlockInventory(Category category, SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe) {
+        super(category, item, recipeType, recipe);
+    }
+}
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/inventory/AbstractVanillaBlockInventory.java:39]: There is no reason for this to be an actual class, this could be handled via an interface. Then again, why do we need to differentiate here anyway? We could use any AContainer for this, couldn't we?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    public HopperListener(@Nonnull SlimefunPlugin plugin) {
+        plugin.getServer().getPluginManager().registerEvents(this, plugin);
+    }
+
+    @EventHandler
+    public void onHopperInsert(InventoryMoveItemEvent e) {
+        Location loc = e.getDestination().getLocation();
+        if (e.getSource().getType() == InventoryType.HOPPER && loc != null && BlockStorage.check(loc.getBlock()) instanceof AbstractVanillaBlockInventory) {
+                e.setCancelled(true);
+        }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java:33]: What about the other direction? Pulling from such an inventory

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java (hunk 14 above)
Comment (CURVX) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/crafting/HopperListener.java]: Sorry, I didn't get what you meant here.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/NotHopperable.java [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.abstractItems.AContainer;
+
+/**
+ * Implement this interface for any {@link AContainer} to prevent
+ * that {@link AContainer} from being hopperable.
+ *
+ * @author CURVX
+ *
+ */
+public interface NotHopperable {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/NotHopperable.java:12]: Needs to extend ItemAttribute

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/HopperListener.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+ * and from {@link AContainer} using a hopper.
+ *
+ * @author CURVX
+ *
+ * @see AutoBrewer
+ * @see AutoDrier
+ * @see ElectricFurnace
+ * @see ElectricIngotPulverizer
+ * @see ElectricOreGrinder
+ * @see ElectricSmeltery
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/HopperListener.java:33]: Should only be `@see NotHopperable`

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Two minor changes still needed.
Then it seems ready code-wise, we may still need to test this in terms of performance.

**Comment by CURVX** [comment]
> Then it seems ready code-wise, we may still need to test this in terms of performance.

How do one test performance? By playing and keeping tabs on timings using spark?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/HopperListener.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+ */
+public class HopperListener implements Listener {
+
+    public HopperListener(@Nonnull SlimefunPlugin plugin) {
+        plugin.getServer().getPluginManager().registerEvents(this, plugin);
+    }
+
+    @EventHandler
+    public void onHopperInsert(InventoryMoveItemEvent e) {
+        Location loc = e.getDestination().getLocation();
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/HopperListener.java:33]: Will this destination inv ever be non-null and an SF item? Custom inventories have `getLocation()` returning null. You're checking if the source is a hopper, therefore, you want this to be SF inv, right?

**Comment by WalshyDev** [comment]
Timing-wise, it's really quick! Getting data from a tile entity is very quick.

I created a fake furnace and fake electric furnace and then ran the InventoryMoveItemEvent 1000 times. This is obviously not doing a full hopper tick but this is a good indication for our code.

Timing data:
lag mode -  showing 3ms
all mode - showing 0.4ms

This event is very quick so there's nothing more we can do on our end. Any lag will be caused by MC itself with their terrible hopper code.

**Review by TheBusyBiscuit — APPROVED** [review]
Alright then :+1:

**Review by WalshyDev — APPROVED** [review]
LGTM
